components/Editor/index.tsx

```diff
@@ -146,8 +148,75 @@ const Editor = ({ readOnly = false }: Props) => {
     return value
   }
 
+  const highlightMnemonic = (value: string) => {
+    if (!codeType) {
+      return value
+    }
+
+    return value
+      .split('\n')
+      .map((line, i) => `<span class='line-number'>${i + 1}</span>${line}`)
+      .join('\n')
+  }
+
   const handleRun = useCallback(() => {
-    if (codeType === CodeType.Bytecode) {
+    if (codeType === CodeType.Mnemonic) {
+      let bytecode = ''
+
+      const lines = code.split('\n')
+      for (let i = 0; i < lines.length; ++i) {
+        const line = lines[i]
+          .replace(/\/\/.*/, '')
+          .trim()
+          .toUpperCase()
+        if (line.length === 0) {
+          continue
+        }
+
+        if (line.startsWith('PUSH')) {
+          const parts = line.split(/\s+/)
+          if (parts.length != 2) {
+            log('Expect PUSH instruction followed by a number: ' + line, 'warn')
+            return
+          }
+
+          const code = opcodes.find((opcode: IOpcode) => {
+            return opcode.name === parts[0]
+          })
+          if (typeof code === 'undefined') {
+            log('Unknown mnemonic: ' + parts[0], 'warn')
+            return
+          }
+
+          const number = parseInt(parts[0].substring(4))
+          const digits = number * 2
+          if (parts[1].length > digits) {
+            log('Number should have at most ' + digits + ' digits: ' + line)
+            return
+          }
+
+          // TODO number checks
+
+          bytecode += code.code
+          bytecode =
+            bytecode.padEnd(bytecode.length + digits - parts[1].length, '0') +
+            parts[1]
+        } else {
+          const code = opcodes.find((opcode: IOpcode) => {
+            return opcode.name === line
+          })
+          if (typeof code === 'undefined') {
+            log('Unknown mnemonic: ' + line, 'warn')
+            return
+          }
+
+          bytecode += code.code
+        }
+      }
```

Comment:
Would be nice to extract it into util, for example in `util/compiler.ts`, which would run these and throw an error. Then this component could catch it and log to UI.

Reply:
I was hoping you could rewrite it, because I assume it is not coded in the best typescript way?